File: stream.py
import threading
import cv2
import numpy as np


class stream(threading.Thread):
    def __init__(self, src):
        threading.Thread.__init__(self)
        self.stream = cv2.VideoCapture(src)
        self.fps = self.stream.get(cv2.CAP_PROP_FPS)
        self.height = self.stream.get(cv2.CAP_PROP_FRAME_HEIGHT)
        self.width = self.stream.get(cv2.CAP_PROP_FRAME_WIDTH)
        self.frameCount = self.stream.get(cv2.CAP_PROP_FRAME_COUNT)
        self.afps = 0
        self.normalFrame2 = None
        self.left_points = []
        self.right_points = []
        self.normalFrame = None
        self.frameToShow = None
        self.locker = True
        self.streamStatus = True

    def run(self):
        while True:
            (grabbed, self.normalFrame) = self.stream.read()
            if not grabbed:
                break
            self.frameToShow = self.normalFrame.copy()
            self.normalFrame2 = self.normalFrame.copy()
            self.draw()
            if cv2.waitKey(int(self.fps)) & 0xFF == ord('q'):
                break
            cv2.imshow('video', self.frameToShow)
        self.stream.release()
        cv2.destroyAllWindows()
        self.streamStatus = False
        exit(0)

    def getStreamStatus(self):
        return self.streamStatus

    def readFrame(self):
        if self.normalFrame is not None:
            return self.normalFrame
        else:
            return None

    def getInfo(self):
        return self.fps, self.width, self.height, self.frameCount

    def setInfo(self, lps, rps, afps):
        self.locker = False
        if len(lps) != 0:
            self.left_points = lps
        if len(rps) != 0:
            self.right_points = rps
        self.afps = afps
        self.locker = True

    def draw(self):
        if self.locker:
            if len(self.right_points) == 0 or len(self.left_points) == 0:
                return
            bottom_points = list()
            top_points = list()
            if len(self.right_points) != 0:
                top_points.append(self.right_points[0])
                bottom_points.append(self.right_points[0])
            if len(self.left_points) != 0:
                top_points.append(self.left_points[0])
                bottom_points.append(self.left_points[0])
            top_points = np.array(top_points, dtype="int32")
            bottom_points = np.array(bottom_points, dtype="int32")
            cv2.polylines(self.normalFrame2, np.int32([top_points]), False, (255, 127, 127), 2, cv2.LINE_AA)
            cv2.polylines(self.normalFrame2, np.int32([bottom_points]), False, (255, 127, 127), 2, cv2.LINE_AA)
            cv2.polylines(self.normalFrame2, np.int32([self.left_points]), False, (255, 127, 127), 2, cv2.LINE_AA)
            cv2.polylines(self.normalFrame2, np.int32([self.right_points]), False, (255, 127, 127), 2, cv2.LINE_AA)
            mask = cv2.inRange(self.normalFrame2, (255, 127, 127), (255, 127, 127))
            im2, contours, hierarchy = cv2.findContours(mask, 1, 2)
            if len(contours) == 0:
                contours = self.prev_contours
            if len(contours) != 0:
                cv2.fillConvexPoly(self.normalFrame2, contours[0], (0, 255, 0))
            cv2.addWeighted(self.normalFrame2, 0.5, self.frameToShow, 1 - 0.5, 0, self.frameToShow)
            self.prev_contours = contours

File: stream.cpp
#include "stream.h"

Stream::Stream(int &argc, char **argv) :
    QApplication(argc, argv),
    width(0), height(0), fps(0)
{
    log.openFile(qApp->applicationDirPath() + "/../logger/logFiles/log_stream.txt", QIODevice::WriteOnly);
    log.write("----------------------------------------------------------");
    log.write("------------------------NEW RUN---------------------------");
    log.write("----------------------------------------------------------");
    timer = new QTimer(this);
    timer->setTimerType(Qt::PreciseTimer);
    connect(timer, SIGNAL(timeout()), this, SLOT(loopOverFrames()));
    updateDataLock = true;
    frameCount = 0;
    colorRange = {0, 0, 255};
    ifGUI = new QDBusInterface("com.stage.gui", "/", "com.stage.gui", bus, this);
    ifMaster = new QDBusInterface("com.stage.master", "/", "com.stage.master", bus2, this);
    ifDetection = new QDBusInterface("com.stage.detection", "/", "com.stage.detection", bus3, this);
    QDBusReply<QString> key = ifMaster->call("getSTREAMGUIKEY");
    sm.setKey(key.value());
    sm.attach(QSharedMemory::ReadWrite);
    QDBusReply<QString> key2 = ifMaster->call("getSTREAMDETECTIONKEY");
    sm2.setKey(key2.value());
    sm2.attach(QSharedMemory::ReadWrite);
}

void Stream::changeStreamInSource(QString source)
{
    reInitStream();
    streamInSource = source;
    if (!cap.open(source.toStdString()))
    {
        log.write("[STREAM] Cannot open video streaming source - exiting..");
        exit(1);
    }
    fps = static_cast<int> (cap.get(cv::CAP_PROP_FPS));
    int frameCount = static_cast<int> (cap.get(cv::CAP_PROP_FRAME_COUNT));
    log.write("[STREAM] opened (" + source + ") - fps: " + QString::number(fps) + " - frameCount: " + QString::number(frameCount/2));
}


void Stream::reInitStream()
{
    if (cap.isOpened())
        cap.release();
    timer->stop();
    ifGUI->call("initViewers");
    log.write("[STREAM] pause/stop streaming");
}

void Stream::loopOverFrames()
{
    t1 = high_resolution_clock::now();
    cap >> inputFrame;
    if (!cap.grab() || inputFrame.cols == 0 || inputFrame.rows == 0)
    {
        reInitStream();
        timer->stop();
        return;
    }
    frameCount++;
    if (inputFrame.cols != 800 || inputFrame.rows != 480)
        resize(inputFrame, inputFrame, Size(800, 480), 0, 0, INTER_LINEAR);
    fsFrame = inputFrame.clone();
    uchar *rawData = fsFrame.data;
    while (!sm.lock());
    sharedData *sData = (sharedData*) sm.data();
    memcpy(sData->rawImg, rawData, FRAME_SIZE);
    sm.unlock();
    ifGUI->call("showFrames");
    ipmTransform();
    uchar *ipmData = ipmFrame.data;
    while (!sm2.lock());
    sharedData2 *sData2 = (sharedData2*) sm2.data();
    memcpy(sData2->ipmData, ipmData, IPM_FRAME_SIZE);
    sm2.unlock();
    ifDetection->call("detect");
    t2 = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(t2 - t1).count();
    timer->setInterval(((1000 / fps) > duration) ? ((1000 / fps) - duration) : 0);
    log.write("[STREAM] frame no. " + QString::number(frameCount) + " - timer interval: " + QString::number(timer->interval()));
}

void Stream::pauseStream()
{
    log.write("[STREAM] pause/stop streaming");
    timer->stop();
}

void Stream::startStream()
{
    log.write("[STREAM] start/continue streaming");
    timer->start(static_cast<int> (1000/fps));
}

void Stream::drawFinalRGB()
{
    Mat processed = fsFrame.clone();
    Mat mask;
    vector<vector<Point>> contours;
    topPts.clear();
    bottomPts.clear();
    topPts.push_back(rightPts.at(0));
    bottomPts.push_back(rightPts.at(rightPts.size()-1));
    topPts.push_back(leftPts.at(0));
    bottomPts.push_back(leftPts.at(leftPts.size()-1));
    polylines(processed, topPts, false, Scalar(0, 0, 255), 2, LINE_8);
    polylines(processed, bottomPts, false, Scalar(0, 0, 255), 2, LINE_8);
    polylines(processed, leftPts, false, Scalar(0, 0, 255), 2, LINE_8);
    polylines(processed, rightPts, false, Scalar(0, 0, 255), 2, LINE_8);
    polylines(fsFrame, leftPts, false, leftColor, 3, LINE_8);
    polylines(fsFrame, rightPts, false, rightColor, 3, LINE_8);
    inRange(processed, colorRange, colorRange, mask);
    findContours(mask, contours, CV_RETR_TREE, CV_CHAIN_APPROX_SIMPLE);
    if (contours.empty())
        contours = prevContours;
    if (contours.size())
    {
        fillConvexPoly(processed, contours[0], Scalar(200, 0, 0), LINE_AA);
        addWeighted(processed, 0.5, fsFrame, 1 - 0.5, 0, fsFrame);
        prevContours = contours;
    }
}

void Stream::ipmTransform()
{
    warpPerspective(inputFrame, ipmFrame, transformHomography, Size(frameHeight, frameWidth));
}

void Stream::setVideoPts(QString videoName)
{
    Mat inputPts;
    Mat dstPts;
    dstPts = (Mat_<double>(4,2) << 0, 0, frameHeight, 0, frameHeight, frameWidth, 0, frameWidth);
    if (videoName == "youtube_video1")
        inputPts = (Mat_<double>(4, 2) << 325, 280, 485, 280, 748, 478, 120, 478);
    else if (videoName == "youtube_video2" || videoName == "youtube_video3")
        inputPts = (Mat_<double>(4, 2) << 340, 280, 520, 280, 770, 478, 120, 478);
    else if (videoName == "youtube_video4")
        inputPts = (Mat_<double>(4, 2) << 310, 280, 445, 280, 615, 390, 200, 390);
    else if (videoName == "youtube_video5")
        inputPts = (Mat_<double>(4, 2) << 365, 300, 490, 300, 700, 440, 210, 440);
    else if (videoName == "youtube_video6")
        inputPts = (Mat_<double>(4, 2) << 380, 220, 480, 220, 610, 350, 235, 350);
    else if (videoName.startsWith("kitti_video"))
        inputPts = (Mat_<double>(4, 2) << 330, 300, 460, 300, 600, 478, 200, 478);
    else if (videoName.startsWith("udacity_video"))
        inputPts = (Mat_<double>(4, 2) << 340, 310, 475, 310, 740, 430, 120, 430);
    inputPts.convertTo(inputPts, CV_32F);
    dstPts.convertTo(dstPts, CV_32F);
    transformHomography = getPerspectiveTransform(inputPts, dstPts);
}

Stream::~Stream()
{
}
